### PR TITLE
[CR] Engine Starting Overhaul

### PR DESCRIPTION
**Pull request by vickio**
Based on my extensive real life experience of owning old, poorly maintained vehicles and having them fail to start at the most inopportune times, I’ve overhauled engine starting to accomplish these primary things:
1.  Sometimes damaged gas/diesel engines fail to start
2.  Engines take time to start and make noise. More if they’re messed up.
3.  It takes more battery to try to start a damaged gas/diesel engine
4.  Damaged gas engines backfire

This creates some nice drama, like having the enemies you’re fleeing from get closer and closer as you desperately crank the beater you’ve carelessly driven through fences and bushes, hoping it will start before they reach you. It also creates some realistic vehicle parts interactions, like draining a car’s battery by trying for too long to start a damaged engine.

**[CR]** There are quite a few factors that I had to come up with numbers for (time to start, loudness, failure chances, etc). I’d like someone with more game dev experience to see if the numbers balance right. And of course you can pick apart my coding too.

Here are some explanations about the changes in this PR:
-   For the whole vehicle to start successfully, all enabled engines must start. It’s all or nothing. It already works that way, but with damaged engines failing to start, this comes into play more. If you have two enabled engines, one perfect and one almost broken, they both must start for you to drive the vehicle (though you can still disable the damaged engine if you’re desperate). I think this is reasonable, but not ideal. To properly handle some engines failing and others succeeding, the multi-engine code would need more work. Currently it doesn't distinguish between a running engine and an enabled engine individually.
-   For the same reason, changing the engine configuration ("Control individual engines”) on a running vehicle will shut off all engines and attempt to restart with the new configuration. This is because I want to make sure starting an individual engine through that interface works the same as a normal vehicle start. I’d like to improve this in the future, but for now, shutting down and restarting the running engines seems like a reasonable workaround to me.
-   Failing to start a vehicle releases the controls if the vehicle is not moving. This was more natural in my testing because if an engine is failing to start due to damage, you usually want to immediately try again rather than having to deliberately “stop driving” in between each attempt.
-   Running damaged gas/diesel engines now have less of a power penalty than other engine types. Specifically, a gas/diesel engine will provide 25% power at worst, rather than 0% as before. I did this for balance because damaged electric/plasma engines never fail to start.
-   I improved the messages about vehicle start failures to give clues about which engine failed and why and to handle multi-engine vehicles better.
-   In cold weather, diesel engines take a little extra time and power to start and are more likely to fail to start when damaged. Believe me, I know.
-   Renamed `start_engine()` to `start_engines()`. It’s only called from a couple of places and this seemed clearer.
-   Created new `start_engine(e)` function for starting an individual engine. This is not necessary at the moment, but if the multi-engine code is improved, this would allow for individual engines to be started properly while the vehicle is running.
-   Removed `msg_start_engine_fail()` since the new engine starting code provides its own messages for start failures. The one thing using it was the engine failure code (engine dies while you’re driving), so I moved its logic there. Really it wasn’t named properly anyway, so we could bring it back into a function called `msg_engine_fail()` if needed.

### CODE REVIEW

**Comment by tivec** [comment]
I haven't read through the code in detail yet, but I **really** like the idea. It gives a bit more "story" to the game, when you're trying to get the vehicle started with a zombie horde. in the area. 

I think it would be a good plan to add a activity for when you are starting the engine. That way, you get a warning when a zombie climbs through your vehicle window. Especially since the start up time can be a very long time and the player can end up with a lot of negative moves.

**Comment by DavidKeaton** [comment]
I second all the things you've added. Love the diesel in cold weather bit. Takes a while for primers to warm up. ;-) How about not only draining batteries, it also has a chance to flood the engine based on driving/mechanics skill?

**Comment by Lain-** [comment]
Looking forward to this, a promp would be nice though for those times when a zombie would sneak up on you and get several free hits in because the player is too busy starting the engine. If the average time to start an engine for those situations is around 4-5 turns then a per-turn confirmation is not bad I think.

**Comment by DavidKeaton** [comment]
@Lain- One could even just use `player_activity` as well! :-)

**Comment by Rivet-the-Zombie** [comment]
Very cool!

File: src/vehicle.cpp [hunk 1]
```diff
@@ -634,9 +633,21 @@ void vehicle::control_engines() {
         cruise_velocity = safe_vel;
     }
 
-    //if an engine isn't running, and player is in control, need to start engine.
-    if (g->u.controlling_vehicle && !engine_on) {
-        start_engine();
+    // Restart engines with new configuration
+    if( dirty ) {
```
Comment (bkentel) [inline src/vehicle.cpp]: You could flip this condition to lose a level of indentation, and you could combine some of the below conditions I think.

```cpp
if (!dirty) {
    return;
}

if (!engine_on) {
    add_msg( _("You change the %s's engine configuration."), name.c_str() );
    return;
}

add_msg( _("You turn off the %s's engines to change their configurations."), name.c_str() );
if (!g->u.controlling_vehicle) {
    return;
}

if (!start_engines() && !velocity) {
    g->u.controlling_vehicle = false;
    add_msg( _("You let go of the controls.") );
}
```

File: src/vehicle.cpp [hunk 2]
```diff
@@ -1316,41 +1310,108 @@ void vehicle::use_controls()
...
+    const vpart_info &einfo = part_info( engines[e] );
+
+    if( !fuel_left( einfo.fuel_type ) ) {
+        add_msg( _("The %s is out of %s."), einfo.name.c_str(), ammo_name( einfo.fuel_type ).c_str() );
+        return false;
+    }
+
+    const bool combustion = einfo.fuel_type == fuel_type_gasoline ||
+        einfo.fuel_type == fuel_type_diesel;
+    const double dmg = 1 - (parts[engines[e]].hp / einfo.durability);
```
Comment (bkentel) [inline src/vehicle.cpp]: Why a `double` if you're doing integer division? You have to cast one of the involved variables if you want a floating point result. This issue occurs in a number of other places as well.

File: src/vehicle.cpp [hunk 3]
```diff
@@ -1316,41 +1310,108 @@ void vehicle::use_controls()
...
+        const point p = parts[engines[e]].mount;
+        int gx, gy;
+        coord_translate( p.x, p.y, gx, gy );
+        sounds::ambient_sound( global_x() + gx, global_y() + gy, start_time / 10, "" );
+    }
+
+    if( combustion ) {
+        // Small engines can be started without a battery (pull start or kick start)
+        if( engine_power >= 50 ) {
+            const int penalty = ((engine_power / 2) * dmg) + ((engine_power / 5) * cold_factor);
```
Comment (bkentel) [inline src/vehicle.cpp]: With small integers you should do the multiply before the divide to avoid truncation: `(1 / 2) * 10` = 0 vs `(1 * 10) / 2 = 5`

**Comment by kevingranade** [comment]
Cool ideas and based on the description (thanks for the thorough description BTW) you seem to have thought this through quite well, I agree with all your points.

**Comment by DavidKeaton** [comment]
Do keep it up mate, I like this PR a lot.

File: src/vehicle.cpp [hunk 4]
```diff
@@ -1316,41 +1310,108 @@ void vehicle::use_controls()
...
+}
+
+void vehicle::backfire( const int e )
+{
+    const int power = part_power( engines[e], true );
+    const point p = parts[engines[e]].mount;
+    int gx, gy;
+    coord_translate( p.x, p.y, gx, gy );
+
+    sounds::ambient_sound( global_x() + gx, global_y() + gy, 40 + (power / 30), "BANG!" );
```
Comment (BevapDin) [inline src/vehicle.cpp]: You can take the coordinates from the vehicle part directly without having them translated (`precalc[0]` is the location of the part in map coordinates, still relative to the vehicles position, but they already include the vehicle rotation):
```C++
point const pos = global_pos() + parts[p].precalc[0];
```

It's done similar (but with separate x,y values) in `fire_turret_internal`:
```C++
bool vehicle::fire_turret_internal (int p, const itype &gun, const itype &ammo, long &charges, const std::string &extra_sound)
{
    int x = global_x() + parts[p].precalc[0].x;
    int y = global_y() + parts[p].precalc[0].y;
    ...
    sounds::sound(x, y, 20, extra_sound);
```

File: src/vehicle.cpp [hunk 5]
```diff
@@ -1316,41 +1310,108 @@ void vehicle::use_controls()
...
-    if (failed_start) {
-        add_msg (_("The %s's engine fails to start."), name.c_str());
-    } else {
-        if (!muscle_powered) {
-            add_msg(_("The %s's engine starts up."), name.c_str());
+    if( attempted == 0 ) {
+        add_msg( _("The %s has no engine."), name.c_str() );
+    } else if( not_muscle > 0 ) {
+        if( started == attempted ) {
+            add_msg( ngettext("The %s's engine starts up.", "The %s's engines start up.", not_muscle), name.c_str() );
```
Comment (BevapDin) [inline src/vehicle.cpp]: The third parameter to `ngettext` is the actual count, in this case probably `attempted`. Same at the line below.

File: src/vehicle.cpp (hunk 1 above)
Comment (vickio) [inline src/vehicle.cpp]: Ok I've reworked that logic to work a bit smoother.

File: src/vehicle.cpp (hunk 2 above)
Comment (vickio) [inline src/vehicle.cpp]: Yeah I switched the scale for some of those variables from 0-100 to 0.0-1.0 but didn't get the casting right. Thanks for catching that, I think it's straightened out now.

File: src/vehicle.cpp (hunk 4 above)
Comment (vickio) [inline src/vehicle.cpp]: Thanks, I knew there must have been a better way!

**Comment by vickio** [comment]
@tivec @Lain- Yeah that makes sense. Right now the longest it would to take to attempt a single engine start is pretty short, less than 2 turns. The start time for a multi-engine vehicle is calculated serially though, which doesn't make sense. In a world where you can have 2 V8 engines in your car, they should both start at the same time.

I think I'll make these changes:

*   Parallel engine starting: slowest engine determines the total vehicle start attempt time
*   Extend the start time a bit so a single attempt could take 3 turns worst case (large diesel on its last leg in very cold weather)
*   Add a player activity so the player gets warned about monsters

Keep in mind, each time you crank on the engine and it fails, you have to interact to try again, which I think makes sense and feels natural in the game. So while 1 attempt might only take 1 or 2 turns, it could take 5 or 6 attempts before the thing actually fires up.

**Comment by KA101** [comment]
Yeah, this looks reasonable.  Let me know when you want it tested.

**Comment by vickio** [comment]
Ok I’ve made the changes as planned. It’s now using `player_activity` and the warnings about monster proximity and attacks feel really good to me and add to the drama of the vehicle starting or not. And of course now you can abort the attempt if you’re getting attacked.

There is one bug I can’t seem to find a solution for. If you get a vehicle going pretty fast, choose stop driving (which turns off the engines), and then start driving again, the engine start attempt doesn’t always happen. This is because I can’t figure out how to properly tell `activity_handlers::start_engines_finish()` which vehicle we’re trying to start. It’s currently passing it the coords where the attempt started, but if the vehicle has moved beyond that point, it can’t locate it and gives up trying to start, leaving you in control of the vehicle without the engine running.

What’s the best way to get the vehicle from the activity handler?

**Comment by BevapDin** [comment]
> What’s the best way to get the vehicle from the activity handler?

Maybe store the vehicles coordinates relative to the player character. The character won't move on its own when a player_activity is handled (e.g. while reading, once can't move because the input is simply ignored). The only way to move the character is through external influence like a hulk pushes it away or the vehicle (with the character in it) moves.

In the first case, the activity would (most likely) be stopped by the player anyway. In the second way, the coordinates still point to the same vehicle.

**Comment by vickio** [comment]
That worked really well except for remote controlling a moving vehicle. So now I'm trying `game::remoteveh()` and falling back to player relative coords for the normal use. `remoteveh()` has some problems of its own, but at least it would handle that case consistently. When trying to remote start the engine of a moving vehicle using "Execute one vehicle action", `remoteveh()` isn't available, and sometimes the player relative coords are too far off, causing it to just silently fail to start the engine. That's kind of an edge case though.

I discovered a crash in `game::remoteveh()` when remote controlling a moving vehicle while inside another moving vehicle, but that's not related to this PR.

**Comment by vickio** [comment]
@KA101 This is ready to be tested.

**Comment by KA101** [comment]
Unfortunately I was busy moving furniture from one location to another and away from my net connection on Friday/Saturday nights.  Apologies for the delay.

**Pull request merged**